google/cloud/storage/object_test.cc

```diff
@@ -190,6 +191,73 @@ TEST_F(ObjectTest, DeleteObjectPermanentFailure) {
       "DeleteObject");
 }
 
+TEST_F(ObjectTest, UpdateObject) {
+  std::string text = R"""({
+      "bucket": "test-bucket-name",
+      "contentDisposition": "new-disposition",
+      "contentLanguage": "new-language",
+      "contentType": "application/octet-stream",
+      "crc32c": "d1e2f3",
+      "etag": "XYZ=",
+      "generation": "12345",
+      "id": "test-bucket-name/test-object-name/12345",
+      "kind": "storage#object",
+      "md5Hash": "xa1b2c3==",
+      "mediaLink": "https://www.googleapis.com/download/storage/v1/b/test-bucket-name/o/test-object-name?generation=12345&alt=media",
+      "metageneration": "4",
+      "name": "test-object-name",
+      "selfLink": "https://www.googleapis.com/storage/v1/b/test-bucket-name/o/test-object-name",
+      "size": 1024,
+      "storageClass": "STANDARD",
+      "timeCreated": "2018-05-19T19:31:14Z",
+      "timeDeleted": "2018-05-19T19:32:24Z",
+      "timeStorageClassUpdated": "2018-05-19T19:31:34Z",
+      "updated": "2018-05-19T19:31:24Z"
+})""";
+  auto expected = ObjectMetadata::ParseFromString(text);
+
+  EXPECT_CALL(*mock, UpdateObject(_))
+      .WillOnce(Return(std::make_pair(TransientError(), ObjectMetadata{})))
+      .WillOnce(Invoke([&expected](internal::UpdateObjectRequest const& r) {
+        EXPECT_EQ("test-bucket-name", r.bucket_name());
+        EXPECT_EQ("test-object-name", r.object_name());
+        EXPECT_THAT(r.json_payload(), HasSubstr("new-disposition"));
+        EXPECT_THAT(r.json_payload(), HasSubstr("new-language"));
+        return std::make_pair(Status(), expected);
+      }));
+  Client client{std::shared_ptr<internal::RawClient>(mock),
+                LimitedErrorCountRetryPolicy(2)};
+
+  auto actual =
+      client.UpdateObject("test-bucket-name", "test-object-name",
+                          ObjectMetadata()
+                              .set_content_disposition("new-disposition")
+                              .set_content_language("new-language"));
```

Comment:
Probably worth adding in checks for the other settable fields too (cacheControl, contentEncoding, contentType, and metadata -- acl seems like a bit more work, but might also be worth updating-and-checking).